### ISSUE TITLE
[TW-506] no garbage collection with CmdAdd

### ISSUE THREAD

**Issue by taskwarrior** [GothenburgBitFactory/taskwarrior#541]
_Scott Kostyshak on 2012-08-05T06:03:05Z says:_ 

 From what I understand, garbage collection should be done before displaying any task IDs to the user. When (at)task add@ is run, as long as (at)context.verbose ("new-id")@ is true, the following string is output to the user: "Created task <ID>". To see why this is a problem, consider the following sequence of commands:
<pre>
scott(at)pango:~$ task add go for a run
Created task 1.
scott(at)pango:~$ task 1 delete
Permanently delete task 1 'go for a run'? (yes/no) yes
Deleting task 1 'go for a run'.
Deleted 1 task.
scott(at)pango:~$ task add eat ice cream
Created task 2.
scott(at)pango:~$ task
[task next]

ID Project Pri Due A Age Urgency Description  
 1                   21s       0 eat ice cream

1 task
</pre>

Note that "Created task 2" should be "Created task 1".

Thus, should CmdAdd be changed so that conditioning on (at)context.verbose ("new-id")@, garbage collection is run? Read-only doesn't need to be checked because I assume that anyone running "add" has write privileges, right?

**Comment by taskwarrior** [comment]
Migrated metadata:
```
Created: 2012-08-05T06:03:05Z
Modified: 2014-02-09T01:57:07Z
```

**Comment by taskwarrior** [comment]
_Wim Schuermann on 2012-08-05T07:08:09Z says:_ 

 The add command is a special case. It does show an ID, but since new tasks are always added to the very end of pending.data this does not pose a problem - all previously shown IDs are still valid, you can think of the output of "add" as an expansion of any previously displayed list of tasks. The newly added task always gets the ID "previous number of tasks in pending.data (including not-yet-garbage-collected-ones) +1", which is a valid ID that can be used for further modifications.

Running garbage collection after an add would invalidate the previously displayed list of task IDs, leading to problems when running, say, a modify command on one of the now possibly reshuffled IDs.

To sum up: This is a special case that is dealt with in a way that makes Taskwarrior dependable. Most people do not even notice that it is a special case because it fits in so nicely. :)

**Comment by taskwarrior** [comment]
_Scott Kostyshak on 2012-08-05T07:54:02Z says:_ 

 Wim Schuermann wrote:
> The newly added task always gets the ID "previous number of tasks in pending.data (including not-yet-garbage-collected-ones) +1", which is a valid ID that can be used for further modifications.

I don't understand this. In the above example, how is 2 a valid ID that can be used for further modifications?

> Running garbage collection after an add would invalidate the previously displayed list of task IDs, leading to problems when running, say, a modify command on one of the now possibly reshuffled IDs.

Ah, I see what you mean here. I could see the common case of someone listing their tasks and then deleting some and adding others. Indeed, a gc would not be wise because after one add (following a delete) their subsequent deletes based on the previously shown list could be misdirected. Thank you for explaining this to me.

**Comment by taskwarrior** [comment]
_Wim Schuermann on 2012-08-05T08:10:15Z says:_ 

 Scott Kostyshak wrote:
> I don't understand this. In the above example, how is 2 a valid ID that can be used for further modifications?

It is a valid ID because the new task is on line 2 of the pending.data, that is all the ID ever means. If you do not run that "next" report (or any report that runs garbage collection, of course) before modifying it, 2 continues to be a valid ID. That is the whole point of only running garbage collection before a report that lists IDs - they remain valid and can be used without constantly running reports to see which ID corresponds to which task now.

The new task continues to be on line 2 of the pending.data until a garbage collection removes the task on line 1, which has been marked as deleted.

If you are still unsure about this, try your above commands, but instead of running the "next" report, modify the task with ID 2 first. Then look at the results.

**Comment by taskwarrior** [comment]
_Scott Kostyshak on 2012-08-05T08:31:03Z says:_ 

 Wim Schuermann wrote:
> Scott Kostyshak wrote:
> > I don't understand this. In the above example, how is 2 a valid ID that can be used for further modifications?
 
> If you are still unsure about this, try your above commands, but instead of running the "next" report, modify the task with ID 2 first. Then look at the results.

Ah, this is where I got confused. I ran (at)task 2 edit@ but it looks like I chose the wrong command to test with as the (at)edit@ command runs (at)gc@. Why does (at)edit@ need to run (at)gc@?

I guess this "bug report" is turning into a lesson. Thank you, Wim, for taking the time to explain these things to me.

**Comment by taskwarrior** [comment]
_Louis-Claude Canon on 2012-08-05T21:15:12Z says:_ 

 The gc is run because there is _displays_id is put to true in the constructor of CmdEdit. If put to false, you can edit task 2 as it remains then a valid ID. The gc is then also called for the following commands: any report, calendar, execute, export, ids, shell, sync.

This may be a bug actually as 'task edit' is not supposed to display IDs more than what 'task info' does.
